lxd/instance/drivers/driver_qemu.go

```diff
@@ -956,21 +956,64 @@ func (d *qemu) saveState(monitor *qmp.Monitor) error {
 	return nil
 }
 
-// Start starts the instance.
-func (d *qemu) Start(stateful bool) error {
-	d.logger.Debug("Start started", log.Ctx{"stateful": stateful})
-	defer d.logger.Debug("Start finished", log.Ctx{"stateful": stateful})
-
+// validateStartup checks any constraints that would prevent start up from succeeding under normal circumstances.
+func (d *qemu) validateStartup(stateful bool) error {
 	// Check that we are startable before creating an operation lock, so if the instance is in the
 	// process of stopping we don't prevent the stop hooks from running due to our start operation lock.
 	err := d.isStartableStatusCode(d.statusCode())
 	if err != nil {
 		return err
 	}
 
-	// Check for stateful.
-	if stateful && !shared.IsTrue(d.expandedConfig["migration.stateful"]) {
-		return fmt.Errorf("Stateful start requires migration.stateful to be set to true")
+	if stateful {
```

Comment:
Sorry, only got to close review now, so this is wrong.

We need this test to be done whenever `migration.stateful` is `true`. Not only when `Start()` is passed `true` for stateful.

That's because `stateful` will only be `true` if the user is asking LXD to restore a stateful state, which is AFTER, we've dumped the state. As the validation we're adding is to make sure that the state can be dumped in the first place, it must always be done, not just when stateful is true.

Reply:
Ah ok that makes sense :+1: